File: apps/api/routers/chats.py
"""
Chats router - handles encrypted chat storage
"""

from typing import Optional, List
from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select

from database import get_db
from models.user import User
from models.chat import Chat
from routers.deps import get_current_user

router = APIRouter()


class CreateChatRequest(BaseModel):
    id: Optional[str] = None
    encrypted_chat_key: str
    chat_key_nonce: str
    encrypted_title: str
    title_nonce: str
    encrypted_chat: str
    chat_nonce: str
    folder_id: Optional[str] = None


class UpdateChatRequest(BaseModel):
    encrypted_title: Optional[str] = None
    title_nonce: Optional[str] = None
    encrypted_chat: Optional[str] = None
    chat_nonce: Optional[str] = None
    folder_id: Optional[str] = None
    pinned: Optional[bool] = None
    archived: Optional[bool] = None


class ChatResponse(BaseModel):
    id: str
    user_id: str
    encrypted_chat_key: str
    chat_key_nonce: str
    encrypted_title: str
    title_nonce: str
    encrypted_chat: str
    chat_nonce: str
    folder_id: Optional[str] = None
    pinned: bool = False
    archived: bool = False
    created_at: int
    updated_at: int


class ChatListItemResponse(BaseModel):
    id: str
    user_id: str
    encrypted_title: str
    title_nonce: str
    encrypted_chat_key: str
    chat_key_nonce: str
    folder_id: Optional[str] = None
    pinned: bool = False
    archived: bool = False
    created_at: int
    updated_at: int


@router.get("", response_model=List[ChatListItemResponse])
async def list_chats(
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    result = await db.execute(
        select(Chat)
        .where(Chat.user_id == user.id)
        .order_by(Chat.updated_at.desc())
    )
    chats = result.scalars().all()

    return [
        ChatListItemResponse(
            id=chat.id,
            user_id=chat.user_id,
            encrypted_title=chat.encrypted_title or "",
            title_nonce=chat.title_nonce or "",
            encrypted_chat_key=chat.encrypted_chat_key or "",
            chat_key_nonce=chat.chat_key_nonce or "",
            folder_id=chat.folder_id,
            pinned=chat.pinned or False,
            archived=chat.archived or False,
            created_at=int(chat.created_at.timestamp()),
            updated_at=int(chat.updated_at.timestamp())
        )
        for chat in chats
    ]


@router.get("/{chat_id}", response_model=ChatResponse)
async def get_chat(
    chat_id: str,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    result = await db.execute(
        select(Chat).where(Chat.id == chat_id, Chat.user_id == user.id)
    )
    chat = result.scalar_one_or_none()

    if not chat:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Chat not found")

    return ChatResponse(
        id=chat.id,
        user_id=chat.user_id,
        encrypted_chat_key=chat.encrypted_chat_key or "",
        chat_key_nonce=chat.chat_key_nonce or "",
        encrypted_title=chat.encrypted_title or "",
        title_nonce=chat.title_nonce or "",
        encrypted_chat=chat.encrypted_chat or "",
        chat_nonce=chat.chat_nonce or "",
        folder_id=chat.folder_id,
        pinned=chat.pinned or False,
        archived=chat.archived or False,
        created_at=int(chat.created_at.timestamp()),
        updated_at=int(chat.updated_at.timestamp())
    )


@router.post("/new", response_model=ChatResponse)
async def create_chat(
    request: CreateChatRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    import uuid
    chat_id = request.id or str(uuid.uuid4())

    chat = Chat(
        id=chat_id,
        user_id=user.id,
        is_encrypted=True,
        encrypted_chat_key=request.encrypted_chat_key,
        chat_key_nonce=request.chat_key_nonce,
        encrypted_title=request.encrypted_title,
        title_nonce=request.title_nonce,
        encrypted_chat=request.encrypted_chat,
        chat_nonce=request.chat_nonce,
        folder_id=request.folder_id
    )
    db.add(chat)
    await db.commit()
    await db.refresh(chat)

    return ChatResponse(
        id=chat.id,
        user_id=chat.user_id,
        encrypted_chat_key=chat.encrypted_chat_key or "",
        chat_key_nonce=chat.chat_key_nonce or "",
        encrypted_title=chat.encrypted_title or "",
        title_nonce=chat.title_nonce or "",
        encrypted_chat=chat.encrypted_chat or "",
        chat_nonce=chat.chat_nonce or "",
        folder_id=chat.folder_id,
        pinned=chat.pinned or False,
        archived=chat.archived or False,
        created_at=int(chat.created_at.timestamp()),
        updated_at=int(chat.updated_at.timestamp())
    )


@router.put("/{chat_id}", response_model=ChatResponse)
async def update_chat(
    chat_id: str,
    request: UpdateChatRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    result = await db.execute(
        select(Chat).where(Chat.id == chat_id, Chat.user_id == user.id)
    )
    chat = result.scalar_one_or_none()

    if not chat:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Chat not found")

    # Update fields
    if request.encrypted_title is not None:
        chat.encrypted_title = request.encrypted_title
    if request.title_nonce is not None:
        chat.title_nonce = request.title_nonce
    if request.encrypted_chat is not None:
        chat.encrypted_chat = request.encrypted_chat
    if request.chat_nonce is not None:
        chat.chat_nonce = request.chat_nonce
    if request.folder_id is not None:
        chat.folder_id = request.folder_id
    if request.pinned is not None:
        chat.pinned = request.pinned
    if request.archived is not None:
        chat.archived = request.archived

    await db.commit()
    await db.refresh(chat)

    return ChatResponse(
        id=chat.id,
        user_id=chat.user_id,
        encrypted_chat_key=chat.encrypted_chat_key or "",
        chat_key_nonce=chat.chat_key_nonce or "",
        encrypted_title=chat.encrypted_title or "",
        title_nonce=chat.title_nonce or "",
        encrypted_chat=chat.encrypted_chat or "",
        chat_nonce=chat.chat_nonce or "",
        folder_id=chat.folder_id,
        pinned=chat.pinned or False,
        archived=chat.archived or False,
        created_at=int(chat.created_at.timestamp()),
        updated_at=int(chat.updated_at.timestamp())
    )


@router.delete("/{chat_id}")
async def delete_chat(
    chat_id: str,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    result = await db.execute(
        select(Chat).where(Chat.id == chat_id, Chat.user_id == user.id)
    )
    chat = result.scalar_one_or_none()
    
    if not chat:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Chat not found")
    
    await db.delete(chat)
    await db.commit()
    
    return {"success": True}

File: apps/server/src/trpc/routers/chats.ts
import { z } from "zod";
import { router, protectedProcedure } from "../trpc";
import { db, chats } from "../../db/client";
import { eq, and, desc } from "drizzle-orm";
import { TRPCError } from "@trpc/server";
import { broadcastToUser } from "../../websocket";

export const chatsRouter = router({
  list: protectedProcedure.query(async ({ ctx }) => {
    const result = await db
      .select()
      .from(chats)
      .where(eq(chats.userId, ctx.user.id))
      .orderBy(desc(chats.updatedAt));

    return result.map((chat) => ({
      id: chat.id,
      userId: chat.userId,
      isEncrypted: chat.isEncrypted,
      encryptedChatKey: chat.encryptedChatKey ?? "",
      chatKeyNonce: chat.chatKeyNonce ?? "",
      encryptedTitle: chat.encryptedTitle ?? "",
      titleNonce: chat.titleNonce ?? "",
      folderId: chat.folderId,
      pinned: chat.pinned,
      archived: chat.archived,
      createdAt: chat.createdAt.getTime(),
      updatedAt: chat.updatedAt.getTime(),
    }));
  }),

  get: protectedProcedure
    .input(z.object({ chatId: z.string().uuid() }))
    .query(async ({ ctx, input }) => {
      const [chat] = await db
        .select()
        .from(chats)
        .where(and(eq(chats.id, input.chatId), eq(chats.userId, ctx.user.id)));

      if (!chat) {
        throw new TRPCError({ code: "NOT_FOUND", message: "Chat not found" });
      }

      return {
        id: chat.id,
        userId: chat.userId,
        isEncrypted: chat.isEncrypted,
        encryptedChatKey: chat.encryptedChatKey ?? "",
        chatKeyNonce: chat.chatKeyNonce ?? "",
        encryptedTitle: chat.encryptedTitle ?? "",
        titleNonce: chat.titleNonce ?? "",
        encryptedChat: chat.encryptedChat ?? "",
        chatNonce: chat.chatNonce ?? "",
        folderId: chat.folderId,
        pinned: chat.pinned,
        archived: chat.archived,
        createdAt: chat.createdAt.getTime(),
        updatedAt: chat.updatedAt.getTime(),
      };
    }),

  create: protectedProcedure
    .input(
      z.object({
        encryptedChatKey: z.string(),
        chatKeyNonce: z.string(),
        encryptedTitle: z.string(),
        titleNonce: z.string(),
        encryptedChat: z.string(),
        chatNonce: z.string(),
        folderId: z.string().uuid().optional(),
        id: z.string().uuid().optional(),
      })
    )
    .mutation(async ({ ctx, input }) => {
      const [chat] = await db
        .insert(chats)
        .values({
          ...(input.id ? { id: input.id } : {}),
          userId: ctx.user.id,
          isEncrypted: true,
          encryptedChatKey: input.encryptedChatKey,
          chatKeyNonce: input.chatKeyNonce,
          encryptedTitle: input.encryptedTitle,
          titleNonce: input.titleNonce,
          encryptedChat: input.encryptedChat,
          chatNonce: input.chatNonce,
          folderId: input.folderId,
          pinned: false,
          archived: false,
        })
        .returning();

      const result = {
        id: chat.id,
        userId: chat.userId,
        isEncrypted: chat.isEncrypted,
        encryptedChatKey: chat.encryptedChatKey ?? "",
        chatKeyNonce: chat.chatKeyNonce ?? "",
        encryptedTitle: chat.encryptedTitle ?? "",
        titleNonce: chat.titleNonce ?? "",
        encryptedChat: chat.encryptedChat ?? "",
        chatNonce: chat.chatNonce ?? "",
        folderId: chat.folderId,
        pinned: chat.pinned,
        archived: chat.archived,
        createdAt: chat.createdAt.getTime(),
        updatedAt: chat.updatedAt.getTime(),
      };

      // Broadcast real-time update
      broadcastToUser(ctx.user.id, "chat:created", result);

      return result;
    }),

  update: protectedProcedure
    .input(
      z.object({
        chatId: z.string().uuid(),
        encryptedTitle: z.string().optional(),
        titleNonce: z.string().optional(),
        encryptedChat: z.string().optional(),
        chatNonce: z.string().optional(),
        folderId: z.string().uuid().nullable().optional(),
        pinned: z.boolean().optional(),
        archived: z.boolean().optional(),
      })
    )
    .mutation(async ({ ctx, input }) => {
      const { chatId, ...updates } = input;

      const [chat] = await db
        .update(chats)
        .set({
          ...updates,
          updatedAt: new Date(),
        })
        .where(and(eq(chats.id, chatId), eq(chats.userId, ctx.user.id)))
        .returning();

      if (!chat) {
        throw new TRPCError({ code: "NOT_FOUND", message: "Chat not found" });
      }

      const result = {
        id: chat.id,
        userId: chat.userId,
        isEncrypted: chat.isEncrypted,
        encryptedChatKey: chat.encryptedChatKey ?? "",
        chatKeyNonce: chat.chatKeyNonce ?? "",
        encryptedTitle: chat.encryptedTitle ?? "",
        titleNonce: chat.titleNonce ?? "",
        encryptedChat: chat.encryptedChat ?? "",
        chatNonce: chat.chatNonce ?? "",
        folderId: chat.folderId,
        pinned: chat.pinned,
        archived: chat.archived,
        createdAt: chat.createdAt.getTime(),
        updatedAt: chat.updatedAt.getTime(),
      };

      // Broadcast real-time update
      broadcastToUser(ctx.user.id, "chat:updated", result);

      return result;
    }),

  remove: protectedProcedure
    .input(z.object({ chatId: z.string().uuid() }))
    .mutation(async ({ ctx, input }) => {
      await db
        .delete(chats)
        .where(and(eq(chats.id, input.chatId), eq(chats.userId, ctx.user.id)));

      // Broadcast real-time update
      broadcastToUser(ctx.user.id, "chat:deleted", { id: input.chatId });

      return { success: true };
    }),
});
